src/psycopmlutils/timeseriesflattener/flattened_dataset.py

```diff
@@ -289,30 +302,225 @@ def _validate_processed_arg_dicts(self, arg_dicts: list):
+        )
+
+        # Expand fallback column
+        df = pd.merge(
+            left=prediction_times_with_uuid_df,
+            right=df,
+            how="left",
+            on=self.pred_time_uuid_col_name,
+            validate="m:1",
         )
 
+        df[full_col_str] = df[full_col_str].fillna(fallback)
+
+        return df
+
+    def _load_most_recent_df_matching_pattern(
+        self,
+        dir: Path,
+        file_pattern: str,
+    ) -> DataFrame:
+        """Load most recent df matching pattern.
+
+        Args:
+            dir (Path): Directory to search
+            file_pattern (str): Pattern to match
+
+        Returns:
+            DataFrame: DataFrame matching pattern
+        """
+        files = list(dir.glob(f"*{file_pattern}*.csv"))
+
+        if len(files) == 0:
+            raise FileNotFoundError(f"No files matching pattern {file_pattern} found")
+
+        most_recent_file = max(files, key=os.path.getctime)
+
+        return pd.read_csv(most_recent_file)
+
+    def _cache_is_hit(
+        self,
+        kwargs_dict: Dict,
+        file_pattern: str,
+        full_col_str: str,
+    ) -> bool:
+        """Check if cache is hit.
+
+        Args:
+            kwargs_dict (Dict): Dictionary of kwargs
+            file_pattern (str): File pattern to match. Looks for *file_pattern* in cache dir.
+            e.g. "*feature_name*_uuids*.csv"
+            full_col_str (str): Full column string. e.g. "feature_name_ahead_interval_days_resolve_multiple_fallback"
+
+        Returns:
+            bool: True if cache is hit, False otherwise
+        """
+        msg = Printer(timestamp=True)
+
+        # Check that file exists
+        file_pattern_hits = list(self.feature_cache_dir.glob(f"*{file_pattern}*.csv"))
+
+        if len(file_pattern_hits) == 0:
+            self.msg.info(f"Cache miss, {file_pattern} didn't exist")
+            return False
+
+        # Check that file contents match expected
+        cache_df = self._load_most_recent_df_matching_pattern(
+            dir=self.feature_cache_dir,
+            file_pattern=file_pattern,
+        )
+
+        generated_df = pd.DataFrame({full_col_str: []})
+
+        # Check that some values in generated_df differ from fallback
+        # Otherwise, comparison to cache is meaningless
+        n_to_generate = 1_000
+
+        while not any(
+            generated_df[full_col_str] != kwargs_dict["fallback"],
+        ):
+            self.msg.info(
+                f"{full_col_str}: Generated_df was all fallback values, regenerating",
+            )
+
+            generated_df = self.flatten_temporal_values_to_df(
+                prediction_times_with_uuid_df=self.df.head(1_000),
+                id_col_name=self.id_col_name,
+                timestamp_col_name=self.timestamp_col_name,
+                pred_time_uuid_col_name=self.pred_time_uuid_col_name,
+                **kwargs_dict,
+            )
+
+            n_to_generate = (
```

Comment:
this will regenerate the same rows as the previous iteration (+ the new ones) each time which potentially take a lot of time

Reply:
Great point! Using `df.sample` instead.